File: Biosphere/Sources/Biosphere/Capabilities/LinearMovement.swift
//
// Pet Therapy.
//

import SwiftUI

public class LinearMovement: Capability {
    
    override public func update(with collisions: Collisions, after time: TimeInterval) {
        guard isEnabled, let body = body else { return }
        body.set(
            origin: body.frame.origin.offset(
                by: movement(after: time)
            )
        )
    }
    
    func movement(after time: TimeInterval) -> CGPoint {
        guard let body = body else { return .zero }
        return CGPoint(
            x: body.direction.dx * body.speed * time,
            y: body.direction.dy * body.speed * time
        )
    }
}

extension Entity {
    
    public var movement: LinearMovement? {
        capability(for: LinearMovement.self)
    }
}

File: Sources/windows/BitTherapyCLR/src/capabilities/linear_movement.cpp
#include "linear_movement.h"

#include <chrono>

#include "../game/game.h"

void LinearMovement::update(std::chrono::milliseconds timeSinceLastUpdate, Entity * entity) {
    auto frameDuration = std::chrono::duration_cast<std::chrono::milliseconds>(timeSinceLastUpdate).count();
    auto offset = entity->direction * entity->speed * frameDuration * 0.001;
    auto updatedFrame = entity->frame.offset(offset);
    entity->frame = updatedFrame;
}